pkg/services/accesscontrol/database/resource_permissions.go

```diff
@@ -329,27 +309,109 @@ func getResourcesPermissions(sess *sqlstore.DBSession, orgID int64, query access
 	builtin := builtinSelect + builtinFrom + where
 	sql := user + "UNION" + team + "UNION" + builtin
 
-	if err := sess.SQL(sql, args...).Find(&result); err != nil {
+	queryResults := make([]flatResourcePermission, 0)
+	if err := sess.SQL(sql, args...).Find(&queryResults); err != nil {
 		return nil, err
 	}
 
 	scopeAll := accesscontrol.GetResourceAllScope(query.Resource)
 	scopeAllIDs := accesscontrol.GetResourceAllIDScope(query.Resource)
-	out := make([]accesscontrol.ResourcePermission, 0, len(result))
 
+	byResource := make(map[string][]flatResourcePermission)
 	// Add resourceIds and generate permissions for `*`, `resource:*` and `resource:id:*`
-	// TODO: handle scope with other key prefixes e.g. `resource:name:*` and `resource:name:name`
 	for _, id := range query.ResourceIDs {
 		scope := accesscontrol.GetResourceScope(query.Resource, id)
-		for _, p := range result {
+		for _, p := range queryResults {
 			if p.Scope == scope || p.Scope == scopeAll || p.Scope == scopeAllIDs || p.Scope == "*" {
 				p.ResourceID = id
-				out = append(out, p)
+				byResource[p.ResourceID] = append(byResource[p.ResourceID], p)
 			}
 		}
 	}
 
-	return out, nil
+	var result []accesscontrol.ResourcePermission
+	for _, permissions := range byResource {
+		users, teams, builtins := groupPermissionsByAssignment(permissions)
+		for _, p := range users {
+			result = append(result, flatPermissionsToResourcePermissions(p)...)
+		}
+		for _, p := range teams {
+			result = append(result, flatPermissionsToResourcePermissions(p)...)
+		}
+		for _, p := range builtins {
+			result = append(result, flatPermissionsToResourcePermissions(p)...)
+		}
```

Comment:
question: why do we group permissions by users, teams and built-in roles just to put them all back in one slice again?